### PR TITLE
[mlir] Added new attributes to the llvm.call op in llvmir target

### PR DESCRIPTION
**Pull request by FMarno**
The new attributes are:
 * convergent
 * no_unwind
 * will_return
 * memory effects

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp [hunk 1]
```diff
@@ -1471,6 +1471,30 @@ LogicalResult ModuleImport::convertInstruction(llvm::Instruction *inst) {
...
+        memEffects.getModRef(llvm::MemoryEffects::Location::Other));
+    auto argMem = convertModRefInfoFromLLVM(
+        memEffects.getModRef(llvm::MemoryEffects::Location::ArgMem));
+    auto inaccessibleMem = convertModRefInfoFromLLVM(
+        memEffects.getModRef(llvm::MemoryEffects::Location::InaccessibleMem));
+    auto memAttr = MemoryEffectsAttr::get(callOp.getContext(), othermem, argMem,
+                                          inaccessibleMem);
+    if (!memAttr.isReadWrite()) {
+      callOp.setMemoryAttr(memAttr);
+    }
```
Comment (Dinistro) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1496]:
```suggestion
    auto othermem = convertModRefInfoFromLLVM(
        memEffects.getModRef(llvm::MemoryEffects::Location::Other));
    auto argMem = convertModRefInfoFromLLVM(
        memEffects.getModRef(llvm::MemoryEffects::Location::ArgMem));
    auto inaccessibleMem = convertModRefInfoFromLLVM(
        memEffects.getModRef(llvm::MemoryEffects::Location::InaccessibleMem));
    auto memAttr = MemoryEffectsAttr::get(callOp.getContext(), othermem, argMem,
                                          inaccessibleMem);
    // Only set the attr when it does not match the default value.
    if (!memAttr.isReadWrite())
      callOp.setMemoryAttr(memAttr);
```

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp [hunk 2]
```diff
@@ -1471,6 +1471,30 @@ LogicalResult ModuleImport::convertInstruction(llvm::Instruction *inst) {
...
+      callOp.setConvergent(true);
+    if (callInst->hasFnAttr(llvm::Attribute::NoUnwind))
+      callOp.setNoUnwind(true);
+    if (callInst->hasFnAttr(llvm::Attribute::WillReturn))
+      callOp.setWillReturn(true);
+
+    // memory effects
+    llvm::MemoryEffects memEffects = callInst->getMemoryEffects();
+    // Only set the attr when it does not match the default value.
+    auto othermem = convertModRefInfoFromLLVM(
```
Comment (Dinistro) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1486]: Nit: Don't use `auto` here, as the type is not given on the RHS.

File: mlir/test/Dialect/LLVMIR/call.mlir
Comment (Dinistro) [inline mlir/test/Dialect/LLVMIR/call.mlir:1]: Don't we already check other call attributes in another test? Would make sense to keep them all together.

File: mlir/test/Target/LLVMIR/Import/call-attributes.ll
Comment (Dinistro) [inline mlir/test/Target/LLVMIR/Import/call-attributes.ll:1]: As above

**Review by Dinistro — COMMENTED** [review]
Thanks for working on this, dropped a few nit comments.

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp [hunk 3]
```diff
@@ -219,6 +219,28 @@ convertOperationImpl(Operation &opInst, llvm::IRBuilderBase &builder,
...
+    if (callOp.getConvergentAttr())
+      call->addFnAttr(llvm::Attribute::Convergent);
+    if (callOp.getNoUnwindAttr())
+      call->addFnAttr(llvm::Attribute::NoUnwind);
+    if (callOp.getWillReturnAttr())
+      call->addFnAttr(llvm::Attribute::WillReturn);
+
+    // memory effects
+    if (callOp.getMemory()) {
+      MemoryEffectsAttr memAttr = callOp.getMemoryAttr();
```
Comment (victor-eds) [inline mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:231]:
```suggestion
    if (MemoryEffectsAttr memAttr = callOp.getMemoryAttr()) {
```

This should work

File: mlir/test/Target/LLVMIR/Import/call-attributes.ll [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+; CHECK: llvm.func @f()
+declare void @f()
+
+; CHECK-LABEL: @call_will_return
+define void @call_will_return() {
+; CHECK: llvm.call @f() {memory = #llvm.memory_effects<other = none, argMem = none, inaccessibleMem = none>}
+  call void @f() memory(none)
+; CHECK: llvm.call @f() {memory = #llvm.memory_effects<other = none, argMem = write, inaccessibleMem = read>}
+  call void @f() memory(none, inaccessiblemem: read, argmem: write)
```
Comment (victor-eds) [inline mlir/test/Target/LLVMIR/Import/call-attributes.ll:47]: Really a nit, but can we sort?

```suggestion
; CHECK: llvm.call @f() {memory = #llvm.memory_effects<other = none, argMem = write, inaccessibleMem = read>}
  call void @f() memory(none, argmem: write, inaccessiblemem: read)
```

**Comment by victor-eds** [comment]
Not really to be addressed in this PR, but IMO this and the `llvm.func` counterpart attribute should be called `memory_effects` as it is the case in LLVM IR. @ftynse are you aware of why the LLVM dialect decided to go with this name instead?

**Comment by Dinistro** [comment]
> Not really to be addressed in this PR, but IMO this and the `llvm.func` counterpart attribute should be called `memory_effects` as it is the case in LLVM IR. @ftynse are you aware of why the LLVM dialect decided to go with this name instead?

I introduced this a while ago, there is no reason to why it's named like this. Feel free to rename for consistency

**Comment by FMarno** [comment]
> > Not really to be addressed in this PR, but IMO this and the `llvm.func` counterpart attribute should be called `memory_effects` as it is the case in LLVM IR. @ftynse are you aware of why the LLVM dialect decided to go with this name instead?
> 
> I introduced this a while ago, there is no reason to why it's named like this. Feel free to rename for consistency

I'd be happy to do this change, but I would rather do it in another PR to minimize the changes here.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp (hunk 1 above)
Comment (FMarno) [reply mlir/lib/Target/LLVMIR/ModuleImport.cpp]: That makes sense. addressed.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp (hunk 2 above)
Comment (FMarno) [reply mlir/lib/Target/LLVMIR/ModuleImport.cpp]: Sure, done

File: mlir/test/Dialect/LLVMIR/call.mlir [hunk 5]
```diff
@@ -0,0 +1,34 @@
+// RUN: mlir-opt -split-input-file -verify-diagnostics %s | mlir-opt | FileCheck %s
+
+module {
```
[the remark this answers is not in the document]
Comment (FMarno) [reply mlir/test/Dialect/LLVMIR/call.mlir]: thanks for the tip, done

File: mlir/test/Dialect/LLVMIR/call.mlir
Comment (FMarno) [reply mlir/test/Dialect/LLVMIR/call.mlir:1]: I've moved the tests as suggested

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp (hunk 3 above)
Comment (FMarno) [reply mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp]: yep, done

**Review by Dinistro — COMMENTED** [review]
Thanks for addressing the comments, LGTM!
Side note: We should really attempt to cleanup some of these test files in the future, as the LLVM dialect tests are a huge mess. I might find some time at some point to do so.

**Review by victor-eds — APPROVED** [review]

**Review by Dinistro — APPROVED** [review]

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp [hunk 6]
```diff
@@ -219,6 +219,27 @@ convertOperationImpl(Operation &opInst, llvm::IRBuilderBase &builder,
...
+    if (callOp.getConvergentAttr())
+      call->addFnAttr(llvm::Attribute::Convergent);
+    if (callOp.getNoUnwindAttr())
+      call->addFnAttr(llvm::Attribute::NoUnwind);
+    if (callOp.getWillReturnAttr())
+      call->addFnAttr(llvm::Attribute::WillReturn);
+
+    // memory effects
+    if (MemoryEffectsAttr memAttr = callOp.getMemoryAttr()) {
+      llvm::MemoryEffects newMemEffects =
```
Comment (gysit) [inline mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:231]:
```suggestion
      llvm::MemoryEffects memEffects =
```
nit: If possible I would avoid new here.

**Review by gysit — APPROVED** [review]
Thanks!

LGTM modulo some ultra nits.

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp [hunk 7]
```diff
@@ -219,6 +219,27 @@ convertOperationImpl(Operation &opInst, llvm::IRBuilderBase &builder,
...
     call->setCallingConv(convertCConvToLLVM(callOp.getCConv()));
     call->setTailCallKind(convertTailCallKindToLLVM(callOp.getTailCallKind()));
+    if (callOp.getConvergentAttr())
+      call->addFnAttr(llvm::Attribute::Convergent);
+    if (callOp.getNoUnwindAttr())
+      call->addFnAttr(llvm::Attribute::NoUnwind);
+    if (callOp.getWillReturnAttr())
+      call->addFnAttr(llvm::Attribute::WillReturn);
+
+    // memory effects
```
[the remark this answers is not in the document]
Comment (FMarno) [reply mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp]: Thanks for your suggestions, I've included them all

**Pull request merged**